src/arch/src/aarch64/cache_info.rs

```diff
@@ -101,9 +118,76 @@ impl CacheType {
     }
 }
 
-impl Default for CacheType {
-    fn default() -> CacheType {
-        CacheType::Unified
+impl CacheInfo {
+    fn sysfs_cache(&self, index: u8, file_name: &str) -> Result<String> {
+        readln_special(&PathBuf::from(format!(
+            "{}/index{}/{}",
+            self.cache_dir.as_path().display(),
+            index,
+            file_name
+        )))
+    }
+
+    pub fn populate(&mut self, index: u8, err_str: &mut String) -> Result<()> {
+        // If the cache level or the type cannot be retrieved we stop the process
+        // of populating the cache levels.
+        match self.sysfs_cache(index, "level") {
+            Ok(level) => {
+                self.level = level
+                    .parse::<u8>()
+                    .map_err(|e| Error::InvalidCacheAttr("level".to_string(), e.to_string()))?;
+            }
+            Err(e) => {
+                // If we cannot read the cache level even for the first level of cache, we will
+                // stop processing anymore cache info and log an error.
+                warn!("Could not read cache level for index {}: {}", index, e);
+                return Err(Error::MissingCacheConfig);
+            }
+        }
+        match self.sysfs_cache(index, "type") {
+            Ok(cache_type) => self.type_ = CacheType::try_from(&cache_type)?,
+            Err(e) => {
+                warn!("Could not read type for cache level {}: {}", self.level, e);
+                return Err(Error::MissingCacheConfig);
+            }
+        }
+
+        if let Ok(shared_cpu_map) = self.sysfs_cache(index, "shared_cpu_map") {
+            self.cpus_per_unit = mask_str2bit_count(shared_cpu_map.trim_end())?;
+        } else {
+            *err_str += "shared cpu map";
+        }
+
+        if let Ok(coherency_line_size) = self.sysfs_cache(index, "coherency_line_size") {
+            self.line_size = Some(coherency_line_size.parse::<u16>().map_err(|e| {
+                Error::InvalidCacheAttr("coherency_line_size".to_string(), e.to_string())
+            })?);
+        } else {
+            *err_str += if *err_str != "" { ", " } else { "" };
```

Comment:
we can simplify this logic by simply adding a comma after each append operation, and pop the last comma right before returning